### ISSUE TITLE
feat(automation): implement `workflow.State`, `workflow.Store`, `Trigger`, and `Hook`

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#1790]
### Test
```
// workflow_test.go
package workflow

func TestWorkflowAdvance(t *testing.T)
func TestWorkflowComplete(t *testing.T)

func TestStoreIntegration(t *testing.T) {
    // create store
    // declare a workflow
    // declare a second workflow
    // s.Create
    // s.Get
    // s.GetDatasetWorkflow
    // s.Create
    // s.List // should have 2
    // s.Update
    // s.Get
    // s.Deploy
    // s.ListDeployed // should be one
    // s.Undeploy
    // s.ListDeployed// should be zero
    // s.Remove
    // s.Get
}
```
### Trigger interface & TriggerConstructor interface
```
package workflow

type Store interface {
    Updater
    Lister
}

type Updater interface {
    Create(did dataset.ID, pid peer.ID, triggers []trigger.Trigger) (*State, error)
}

type Lister interface {
    // List lists all the workflows in the store
    List(ctx context.Context, offset, limit int) ([]*Workflow, error)
    // ListDeployed lists all the deployed workflows in the store
    ListDeployed(ctx context.Context, offset limit int) ([]*Workflow, error)
}

type State struct {
    ...
    triggers []trigger.Trigger
}

func (m *MemStore) Create(did dataset.ID, pid peer.ID, triggers []trigger.Trigger) (*State, error)
func (m *MemStore) List(ctx context.Context, offset, limit int) ([]*Workflow, error)
func (m *MemStore) ListDeployed(ctx context.Context, offset, limit int) ([]*Workflow, error)
```
```
// workflow/trigger.go
package workflow

// Type describes the type of trigger the system listens for
type TriggerType string

// Trigger has all the information needed to determine under what circumstances an `event.ETTriggerWorkflow`
// should be emitted. A Trigger should also know how to "advance" itself once a trigger was... triggered.
type Trigger interface {
    json.Marshaler
    json.Unmarshaler
    ID() string
    Enabled() boolean
    Type() TriggerType
    ToMap() map[string]interface{}
    Advance() error
}

```
### Trigger tests
```
// trigger_test.go
package workflow

type TestTrigger struct {
}

```

### Hook & HookConstructor interface
```
// workflow/hook.go
package workflow

// Type describes the type of hook the system listens for
type HookType string

// Hook has all the information needed to determine how to react to a transform completed event
type Hook interface {
    json.Marshaler
    json.Unmarshaler
    ID() string
    Enabled() boolean
    Type() HookType
    ToMap() map[string]interface{}
    Advance() error
}
```
### Hook tests
```
// hook_test.go
package workflow

type TestHook struct {
}
```

### expand Workflow methods
```
package workflow

// Workflow holds both the instructions on how to run an automation
type Workflow struct { /* ... */ }

// Advance sets the workflow up for a new run
func (w *Workflow) Advance() error

// Complete records any information needed to note at the end of the run
func (w *Workflow) Complete() error

// Store handles the persistence and listing of Workflows. Store implementations must be safe for concurrent use
type Store interface {
	Lister
	// Creates a new undeployed workflow
	Create(wf *Workflow) (*Workflow, error)
	// Updates a workflow 
	Update(wf *Workflow) error
	// Remove removes a workflow from the store
	Remove(id string) error
	// Get a workflow from a workflow id
	Get(id string) (*Workflow, error)
	// Deploy enables the workflow to be run by an orchestrator
	Deploy(id string) error
	// Undeploy disables the workflow from being run by an orchestrator
	Undeploy(id string) error
        // GetDatasetWorkflow fetches a workflow using the dataset id
        // GetDatasetWorkflow(did string) (*Workflow, error)
}

// Lister lists workflows
type Lister interface {
	// List lists all the workflows in the store
	List(ctx context.Context, offset, limit int) ([]*Workflow, error)
	// ListDeployed lists all the deployed workflows
	ListDeployed(ctx context.Context, status Status, offset, limit int) ([]*Workflow, error)
}

var _ Store = (*MemStore)(nil)
```

**Comment by ramfox** [comment]
@b5 potentially controversial but wanted to bring it up. This could not be viable if we have multiple workflows for a single dataset in the future (not sure how this would work), but what if we nixed the workflow id, and used the dataset id as the primary identifier for a workflow?

Or maybe we still have workflow IDs but they aren't the primary way to fetch/store a workflow? Either way, need to add fetching via dataset id as a method on `workflow.Store`, since I can think of multiple instances where this would be a more convenient route to getting workflow.

**Comment by b5** [comment]
> what if we nixed the workflow id, and used the dataset id as the primary identifier for a workflow?

We need workflow IDs b/c workflows & datasets can have a many-to-one relationship. Individual nodes will refuse to create multiple workflows for the same dataset ID, but that can't be coordinated across peers. Using unique IDs for workflows gives us tools to disambiguate in that context. It would be possible to use the combination of `profileID`+`datasetID` to create a unique identifier, but that string is longer than a UUID, and I'm not convinced we have the definition of a `profileID` totally worked out.

> Either way, need to add fetching via dataset id as a method on workflow.Store, since I can think of multiple instances where this would be a more convenient route to getting workflow.

_Completely_ agree with this. It'll be a very common pattern to get a workflow for a given dataset ID. I wouldn't be surprised to see that in practice the order of usage is `store.List`, `store.GetDatasetWorkflow`, `store.Get(id)` . 

To keep these two thoughts consistent any `workflow.Store` implementation MUST reject any attempt to add  a workflow that uses a `dataset.ID` that's already associated with a workflow in the store.

**Comment by ramfox** [comment]
Dope, v helpful ty

**Comment by ramfox** [comment]
We've adjusted the package structure and moved some interfaces around!

We are adding a `trigger.TriggerSourceStore` that returns `trigger.TriggerSource`s aka `Workflow`s, and a `hook.HookSourceStore` that returns `hook.HookSource`s also aka `Workflow`s (based on this comment: https://github.com/qri-io/qri/pull/1797#issuecomment-850549868)

This will now allow us to import `trigger.Trigger` and `hook.Hook` into `workflow`, since we no longer need to import `workflow.Store` into `trigger` and `hook`. This change makes the packages much more logical.

We've also decided to move the `spec` package into `automation`, and have it contain spec for each subsystem.

- automation sub-packages
   - workflow - defines `Workflow` & `Store`
   - trigger - defines `Trigger`, `Watcher`, and their implementations, starting with `RuntimeTrigger` and `RuntimeWatcher` (the latter of which is not yet defined), which are expected to be used in testing environments
   - hook - defines `Hook` & `Completor`, and their implementations, starting with `RuntimeHook` and `RuntimeCompletor` (the latter of which is not yet defined), which are expected to be used in testing environments
   - run - defines a `Run` & `RunStore`
   - spec - defines spec tests for the `workflow.Store`, `trigger.Trigger`, `trigger.Watcher`, `hook.Hook`, `hook.Completor`, `run.Store`
